<a href="https://colab.research.google.com/github/Adlucent/ga4-return-prediction/blob/main/7_Model_Building_and_Evaluation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 7. Model Building

In [ ]:
################################################################################
######################### CHANGE BQ PROJECT NAME BELOW #########################
################################################################################

project_name = 'adl-analytics' #add proj name
region = "US"  # GCP project region

In [ ]:
# If your notebook does not have pandas_gbq you can install it here:
# ! pip install pandas_gbq

In [ ]:
# Google credentials
from google.colab import auth
auth.authenticate_user()

# BigQuery Magics
'''BigQuery magics are used to run BigQuery SQL queries in a python environment.
These queries can also be run in the BigQuery UI '''

from google.cloud import bigquery
from google.cloud.bigquery import magics, Client, QueryJobConfig

magics.context.project = project_name #update project name
client = bigquery.Client(project=magics.context.project)

# Interface between Jupyter and BigQuery
import pandas_gbq

# data processing libraries + ML tools
import pandas as pd
import numpy as np

# ML model evaluation metrics
from sklearn.metrics import (
    explained_variance_score,
    mean_absolute_error,
    mean_squared_error,
    mean_squared_log_error,
    median_absolute_error,
    r2_score
)

# visualization
import matplotlib.pyplot as plt
import seaborn as sns

sns.set()

# suppress notebook warnings
import warnings
warnings.filterwarnings('ignore')

# dataframe formatting
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", None)
pd.set_option("display.float_format", lambda x: "%.3f" % x)

# Build Models
Using BigQuery ML

## Linear Regression Model
See hyperparameter details and model options [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-glm). <BR>
**EXPLAIN HYPERPARAMETERS BEING TUNED**

In [ ]:
%%bigquery lr_model --project $project_name
CREATE OR REPLACE MODEL
  `return_prediction_ga4.lr_model`
OPTIONS
  ( MODEL_TYPE='LINEAR_REG',
    FIT_INTERCEPT = TRUE,
    L1_REG = HPARAM_RANGE(0, 10.0),
    L2_REG = HPARAM_RANGE(0, 10.0),
    MAX_ITERATIONS = 50,
    EARLY_STOP = FALSE,
    DATA_SPLIT_METHOD='CUSTOM',
    DATA_SPLIT_COL='SPLIT_COL',
    NUM_TRIALS = 20,
    HPARAM_TUNING_ALGORITHM = 'VIZIER_DEFAULT',
    HPARAM_TUNING_OBJECTIVES = ['MEAN_ABSOLUTE_ERROR']) AS
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'TRAIN' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'TRAIN'
UNION ALL
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'EVAL' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'VALID'
UNION ALL
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'TEST' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'TEST';

Query is running:   0%|          |

## RandomForest Model
See hyperparameter details and model options [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-random-forest). <br>
**EXPLAIN HYPERPARAMETERS BEING TUNED**

In [ ]:
%%bigquery rf_model --project $project_name
CREATE OR REPLACE MODEL
    `return_prediction_ga4.rf_model`
OPTIONS
  ( MODEL_TYPE='RANDOM_FOREST_REGRESSOR',
    NUM_PARALLEL_TREE = HPARAM_RANGE(2, 200),
    MAX_TREE_DEPTH = HPARAM_RANGE(1, 20),
    MIN_TREE_CHILD_WEIGHT = HPARAM_RANGE(2, 20),
    COLSAMPLE_BYNODE = HPARAM_RANGE(0.25, 0.75),
    SUBSAMPLE = HPARAM_RANGE(0.5, 0.9),
    EARLY_STOP = TRUE,
    MIN_REL_PROGRESS = 0.0001,
    DATA_SPLIT_METHOD='CUSTOM',
    DATA_SPLIT_COL='SPLIT_COL',
    NUM_TRIALS = 50,
    HPARAM_TUNING_ALGORITHM = 'VIZIER_DEFAULT',
    HPARAM_TUNING_OBJECTIVES = ['MEAN_ABSOLUTE_ERROR'] ) AS
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'TRAIN' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'TRAIN'
UNION ALL
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'EVAL' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'VALID'
UNION ALL
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'TEST' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'TEST';

Query is running:   0%|          |

## XGBoost Model
See hyperparameter details and model options [here](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-create-boosted-tree). <br>
**EXPLAIN HYPERPARAMETERS BEING TUNED**

In [ ]:
%%bigquery xgb_model --project $project_name
CREATE OR REPLACE MODEL
    `return_prediction_ga4.xgb_model`
OPTIONS
  ( MODEL_TYPE='BOOSTED_TREE_REGRESSOR',
    BOOSTER_TYPE = HPARAM_CANDIDATES(['GBTREE', 'DART']),
    DART_NORMALIZE_TYPE = HPARAM_CANDIDATES(['TREE', 'FOREST']),
    NUM_PARALLEL_TREE = HPARAM_RANGE(2, 100),
    MAX_TREE_DEPTH = HPARAM_RANGE(1, 20),
    DROPOUT = HPARAM_RANGE(0, 0.5),
    LEARN_RATE = HPARAM_RANGE(0.1, 0.5),
    MIN_TREE_CHILD_WEIGHT = HPARAM_RANGE(2, 20),
    COLSAMPLE_BYNODE = HPARAM_RANGE(0.25, 0.75),
    SUBSAMPLE = HPARAM_RANGE(0.5, 0.9),
    EARLY_STOP = TRUE,
    MIN_REL_PROGRESS = 0.0001,
    MAX_ITERATIONS = 50,
    DATA_SPLIT_METHOD='CUSTOM',
    DATA_SPLIT_COL='SPLIT_COL',
    NUM_TRIALS = 90,
    HPARAM_TUNING_ALGORITHM = 'VIZIER_DEFAULT',
    HPARAM_TUNING_OBJECTIVES = ['MEAN_ABSOLUTE_ERROR'] ) AS
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'TRAIN' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'TRAIN'
UNION ALL
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'EVAL' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'VALID'
UNION ALL
SELECT
    * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
    ecommerce_refund_value_in_usd AS label,
    'TEST' AS split_col
FROM
    `return_prediction_ga4.step_6_merged_scaled`
WHERE
    split = 'TEST';

Query is running:   0%|          |

# Best Models
Compare the best models from each model type we tested, ie. the best hyperparameter combinations based on our evaluation metric, mean absolute error (MAE). <br>
BigQuery automatically tells us which hyperparameter combination performed the best as 'is_optimal' = True.

### Linear Regression

In [ ]:
%%bigquery lr_model_performance --project $project_name
SELECT
  *
FROM
  ML.TRIAL_INFO(MODEL `adl-analytics.return_prediction_ga4.linreg_model`);

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
# all lr model trial performances
lr_model_performance

trial_id                                    hyperparameters  \
0          1                 {'l1_reg': 1e-14, 'l2_reg': 1e-14}   
1          2  {'l1_reg': 1.991669131404525e-11, 'l2_reg': 1....   
2          3  {'l1_reg': 2.8692991443926807e-13, 'l2_reg': 1...   
3          4  {'l1_reg': 1.5818032333206806e-07, 'l2_reg': 1...   
4          5  {'l1_reg': 0.002506451730963958, 'l2_reg': 1e-14}   
5          6      {'l1_reg': 9.99999999999999, 'l2_reg': 1e-14}   
6          7  {'l1_reg': 9.99999999999999, 'l2_reg': 6.30945...   
7          8  {'l1_reg': 9.99999999999999, 'l2_reg': 7.94313...   
8          9  {'l1_reg': 9.99999999999999, 'l2_reg': 9.99999...   
9         10  {'l1_reg': 0.015728685259771075, 'l2_reg': 1.3...   
10        11  {'l1_reg': 9.99999999999999, 'l2_reg': 4.47675...   
11        12  {'l1_reg': 9.99999999999999, 'l2_reg': 1.52181...   
12        13  {'l1_reg': 9.99999999999999, 'l2_reg': 3.45164...   
13        14  {'l1_reg': 9.99999999999999, 'l2_reg': 1.55041...   
14        15  {'l1_reg': 9.99999999999999, 'l2_reg': 7.96993...   
15        16      {'l1_reg': 1e-14, 'l2_reg': 9.99999999999999}   
16        17  {'l1_reg': 9.99999999999999, 'l2_reg': 7.27884...   
17        18  {'l1_reg': 9.999999999999845, 'l2_reg': 1.0679...   
18        19  {'l1_reg': 9.99999999999999, 'l2_reg': 3.10449...   
19        20  {'l1_reg': 1e-14, 'l2_reg': 3.821562980387304e...   

               hparam_tuning_evaluation_metrics  training_loss  eval_loss  \
0   {'mean_absolute_error': 20.464286656654963}       2632.698    888.989   
1    {'mean_absolute_error': 20.46428665665614}       2632.698    888.989   
2   {'mean_absolute_error': 20.464286656654963}       2632.698    888.989   
3   {'mean_absolute_error': 20.464286656585877}       2632.698    888.989   
4     {'mean_absolute_error': 20.4642855618118}       2632.698    888.989   
5    {'mean_absolute_error': 20.45994517691412}       2632.610    888.860   
6   {'mean_absolute_error': 20.459945176951855}       2632.610    888.860   
7    {'mean_absolute_error': 20.45994992343326}       2632.610    888.860   
8    {'mean_absolute_error': 21.05898792354632}       2623.497    912.302   
9     {'mean_absolute_error': 20.4642798671829}       2632.698    888.989   
10  {'mean_absolute_error': 20.459945203665566}       2632.610    888.860   
11   {'mean_absolute_error': 20.45994517691422}       2632.610    888.860   
12   {'mean_absolute_error': 20.45994517691619}       2632.610    888.860   
13  {'mean_absolute_error': 20.459945177840606}       2632.610    888.860   
14  {'mean_absolute_error': 20.459945176914122}       2632.610    888.860   
15  {'mean_absolute_error': 21.063623293725072}       2623.618    912.458   
16  {'mean_absolute_error': 20.459945176914573}       2632.610    888.860   
17  {'mean_absolute_error': 20.459945183295996}       2632.610    888.860   
18   {'mean_absolute_error': 20.45994517709966}       2632.610    888.860   
19  {'mean_absolute_error': 20.464286885108226}       2632.698    888.989   

       status error_message  is_optimal  
0   SUCCEEDED          None       False  
1   SUCCEEDED          None       False  
2   SUCCEEDED          None       False  
3   SUCCEEDED          None       False  
4   SUCCEEDED          None       False  
5   SUCCEEDED          None        True  
6   SUCCEEDED          None       False  
7   SUCCEEDED          None       False  
8   SUCCEEDED          None       False  
9   SUCCEEDED          None       False  
10  SUCCEEDED          None       False  
11  SUCCEEDED          None       False  
12  SUCCEEDED          None       False  
13  SUCCEEDED          None       False  
14  SUCCEEDED          None       False  
15  SUCCEEDED          None       False  
16  SUCCEEDED          None       False  
17  SUCCEEDED          None       False  
18  SUCCEEDED          None       False  
19  SUCCEEDED          None       False

In [ ]:
list(lr_model_performance[lr_model_performance['is_optimal'] == True]['hyperparameters'])

[{'l1_reg': 9.99999999999999, 'l2_reg': 1e-14}]

### RandomForest

In [ ]:
%%bigquery rf_model_performance --project $project_name
SELECT
  *
FROM
  ML.TRIAL_INFO(MODEL `adl-analytics.return_prediction_ga4.rf_model`);

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
# all rf model trial performances
rf_model_performance

trial_id                                    hyperparameters  \
0          1  {'max_tree_depth': 15, 'subsample': 0.8, 'num_...   
1          2  {'max_tree_depth': 12, 'subsample': 0.71609730...   
2          3  {'max_tree_depth': 7, 'subsample': 0.628120700...   
3          4  {'max_tree_depth': 17, 'subsample': 0.78884810...   
4          5  {'max_tree_depth': 15, 'subsample': 0.62010212...   
5          6  {'max_tree_depth': 8, 'subsample': 0.728855685...   
6          7  {'max_tree_depth': 12, 'subsample': 0.73402620...   
7          8  {'max_tree_depth': 12, 'subsample': 0.82409124...   
8          9  {'max_tree_depth': 14, 'subsample': 0.67460912...   
9         10  {'max_tree_depth': 14, 'subsample': 0.81632796...   
10        11  {'max_tree_depth': 15, 'subsample': 0.68228370...   
11        12  {'max_tree_depth': 17, 'subsample': 0.57712177...   
12        13  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
13        14  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
14        15  {'max_tree_depth': 19, 'subsample': 0.5, 'num_...   
15        16  {'max_tree_depth': 16, 'subsample': 0.5, 'num_...   
16        17  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
17        18  {'max_tree_depth': 19, 'subsample': 0.5, 'num_...   
18        19  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
19        20  {'max_tree_depth': 16, 'subsample': 0.5, 'num_...   
20        21  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
21        22  {'max_tree_depth': 20, 'subsample': 0.70612320...   
22        23  {'max_tree_depth': 19, 'subsample': 0.62195172...   
23        24  {'max_tree_depth': 20, 'subsample': 0.55582976...   
24        25  {'max_tree_depth': 20, 'subsample': 0.58842907...   
25        26  {'max_tree_depth': 20, 'subsample': 0.55991769...   
26        27  {'max_tree_depth': 20, 'subsample': 0.58834618...   
27        28  {'max_tree_depth': 20, 'subsample': 0.55050405...   
28        29  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
29        30  {'max_tree_depth': 17, 'subsample': 0.5, 'num_...   
30        31  {'max_tree_depth': 1, 'subsample': 0.9, 'num_p...   
31        32  {'max_tree_depth': 20, 'subsample': 0.53798995...   
32        33  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
33        34  {'max_tree_depth': 20, 'subsample': 0.56063126...   
34        35  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
35        36  {'max_tree_depth': 15, 'subsample': 0.50036298...   
36        37  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
37        38  {'max_tree_depth': 16, 'subsample': 0.5, 'num_...   
38        39  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
39        40  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
40        41  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
41        42  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
42        43  {'max_tree_depth': 1, 'subsample': 0.9, 'num_p...   
43        44  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   
44        45  {'max_tree_depth': 11, 'subsample': 0.5, 'num_...   
45        46  {'max_tree_depth': 18, 'subsample': 0.5, 'num_...   
46        47  {'max_tree_depth': 20, 'subsample': 0.87780606...   
47        48  {'max_tree_depth': 12, 'subsample': 0.5, 'num_...   
48        49  {'max_tree_depth': 11, 'subsample': 0.5, 'num_...   
49        50  {'max_tree_depth': 20, 'subsample': 0.5, 'num_...   

               hparam_tuning_evaluation_metrics  training_loss  eval_loss  \
0   {'mean_absolute_error': 16.297838421311745}         24.327     35.077   
1   {'mean_absolute_error': 16.269731785915337}         31.549     33.719   
2    {'mean_absolute_error': 16.53163167969424}         39.306     33.089   
3   {'mean_absolute_error': 16.342993546641818}         37.124     33.241   
4   {'mean_absolute_error': 16.322093485932378}         28.978     33.350   
5     {'mean_absolute_error': 16.6232696973623}         30.709     34.660   
6   {'mean_absolute_error': 16.334606689598793}

In [ ]:
list(rf_model_performance[rf_model_performance['is_optimal'] == True]['hyperparameters'])

[{'max_tree_depth': 20,
  'subsample': 0.5,
  'num_parallel_tree': 3,
  'min_tree_child_weight': 2,
  'colsample_bynode': 0.25000000000000216}]

### RandomForest

In [ ]:
%%bigquery xgb_model_performance --project $project_name
SELECT
  *
FROM
  ML.TRIAL_INFO(MODEL `adl-analytics.return_prediction_ga4.xgb_model`);

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
# all xgb model trial performances
xgb_model_performance

trial_id                                    hyperparameters  \
0          1  {'learn_rate': 0.3, 'dropout': None, 'max_tree...   
1          2  {'learn_rate': 0.4978354418496759, 'dropout': ...   
2          3  {'learn_rate': 0.3833610902957896, 'dropout': ...   
3          4  {'learn_rate': 0.46779875400958404, 'dropout':...   
4          5  {'learn_rate': 0.43142461413494737, 'dropout':...   
5          6  {'learn_rate': 0.49788202544016025, 'dropout':...   
6          7  {'learn_rate': 0.289677670546588, 'dropout': N...   
7          8  {'learn_rate': 0.4126402492841693, 'dropout': ...   
8          9  {'learn_rate': 0.19269102223466822, 'dropout':...   
9         10  {'learn_rate': 0.2932046182675494, 'dropout': ...   
10        11  {'learn_rate': 0.1, 'dropout': None, 'max_tree...   
11        12  {'learn_rate': 0.1, 'dropout': None, 'max_tree...   
12        13  {'learn_rate': 0.1, 'dropout': None, 'max_tree...   
13        14  {'learn_rate': 0.5, 'dropout': 0.2800554794453...   
14        15  {'learn_rate': 0.1, 'dropout': None, 'max_tree...   
15        16  {'learn_rate': 0.1, 'dropout': None, 'max_tree...   
16        17  {'learn_rate': 0.5, 'dropout': 0.4894819379784...   
17        18  {'learn_rate': 0.40667060298118585, 'dropout':...   
18        19  {'learn_rate': 0.31666065213909, 'dropout': 0....   
19        20  {'learn_rate': 0.2180578336456694, 'dropout': ...   
20        21  {'learn_rate': 0.1167824059528334, 'dropout': ...   
21        22  {'learn_rate': 0.1, 'dropout': 0.3700564511833...   
22        23  {'learn_rate': 0.1, 'dropout': None, 'max_tree...   
23        24  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
24        25  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
25        26  {'learn_rate': 0.1, 'dropout': 0.4184727576554...   
26        27  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
27        28  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
28        29  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
29        30  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
30        31  {'learn_rate': 0.39674629238367787, 'dropout':...   
31        32  {'learn_rate': 0.10004248675884808, 'dropout':...   
32        33  {'learn_rate': 0.10000800201259447, 'dropout':...   
33        34  {'learn_rate': 0.1, 'dropout': 0.4943300991546...   
34        35  {'learn_rate': 0.1, 'dropout': 0.3334777356557...   
35        36  {'learn_rate': 0.1, 'dropout': 0.3301422884582...   
36        37  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
37        38  {'learn_rate': 0.1, 'dropout': 0.4999887073695...   
38        39  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
39        40  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
40        41  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
41        42  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
42        43  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
43        44  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
44        45  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
45        46  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
46        47  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
47        48  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
48        49  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
49        50  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
50        51  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
51        52  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
52        53  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
53        54  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
54        55  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
55        56  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
56        57  {'learn_rate': 0.1, 'dropout': 0.5, 'max_tree_...   
57        58  {'learn_rate': 0.1, 'dropout': 0.2847297399220...   
58        59  {'learn_rate': 0.1, 'dropout': 0.5, '

In [ ]:
list(xgb_model_performance[xgb_model_performance['is_optimal'] == True]['hyperparameters'])

[{'learn_rate': 0.1,
  'dropout': 0.5,
  'max_tree_depth': 20,
  'subsample': 0.5,
  'booster_type': 'dart',
  'num_parallel_tree': 46,
  'dart_normalize_type': 'tree',
  'min_tree_child_weight': 2,
  'colsample_bynode': 0.42820566106224267}]

## Model Selection
These three models with the best hyperparameters for each model type will be evaluated against each other on the validation set to choose the final model. <br>
The model with the best performance on validation will be chosen as the final model. Lastly, we can use our final model to predict on the test set.

**Best model hyperparameters for Linear Regression:** <br>
- L1 regularization: 10, <br>
- L2 regularization: 0 <br>

**Best model hyperparameters for RandomForest Regressor:** <br>
- Maximum tree depth: 20, <br>
- Subsample: 0.5 <br>
- Number of parallel trees: 3 <br>
- Minimum tree child weight: 2 <br>
- Column sample by node: 0.25 <br>

**Best model hyperparameters for RandomForest Regressor:** <br>
- Learning rate: 0.1, <br>
- Dropout: 0.5 <br>
- Maximum tree depth: 20 <br>
- Subsample: 0.5 <br>
- Booster type: 'dart' <br>
- Number of parallel trees: 46 <br>
- Dart normalization type: 'tree' <br>
- Minimum tree child weight: 2 <br>
- Column sample by node: 0.4282 <br>

In [ ]:
%%bigquery lr_model_eval --project $project_name
SELECT
    *
FROM
    ML.EVALUATE(MODEL `adl-analytics.return_prediction_ga4.linreg_model`,
        (
            SELECT
                * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
                ecommerce_refund_value_in_usd AS label,
            FROM
                `return_prediction_ga4.step_6_merged_scaled`
            WHERE
                split = 'VALID'
        ),
            STRUCT(6 AS trial_id));

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
lr_model_eval

trial_id  mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0         6               20.460             888.860                   5.256   

   median_absolute_error  r2_score  explained_variance  
0                 14.782    -0.136               0.041

In [ ]:
%%bigquery rf_model_eval --project $project_name
SELECT
    *
FROM
    ML.EVALUATE(MODEL `adl-analytics.return_prediction_ga4.rf_model`,
        (
            SELECT
                * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
                ecommerce_refund_value_in_usd AS label,
            FROM
                `return_prediction_ga4.step_6_merged_scaled`
            WHERE
                split = 'VALID'
        ),
            STRUCT(33 AS trial_id));

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
rf_model_eval

trial_id  mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0        33               13.897            1025.901                   3.363   

   median_absolute_error  r2_score  explained_variance  
0                  3.491    -0.312              -0.302

In [ ]:
%%bigquery xgb_model_eval --project $project_name
SELECT
    *
FROM
    ML.EVALUATE(MODEL `adl-analytics.return_prediction_ga4.xgb_model`,
        (
            SELECT
                * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
                ecommerce_refund_value_in_usd AS label,
            FROM
                `return_prediction_ga4.step_6_merged_scaled`
            WHERE
                split = 'VALID'
        ),
            STRUCT(55 AS trial_id));

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
xgb_model_eval

trial_id  mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0        55               11.984             859.500                   2.277   

   median_absolute_error  r2_score  explained_variance  
0                  1.217    -0.099               0.025

In [ ]:
best_model_evals = pd.concat([lr_model_eval, rf_model_eval, xgb_model_eval])
best_model_evals['model_type'] = ['linear_regression', 'random_forest', 'xgboost']
best_model_evals.sort_values('mean_absolute_error')

trial_id  mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0        55               11.984             859.500                   2.277   
0        33               13.897            1025.901                   3.363   
0         6               20.460             888.860                   5.256   

   median_absolute_error  r2_score  explained_variance         model_type  
0                  1.217    -0.099               0.025            xgboost  
0                  3.491    -0.312              -0.302      random_forest  
0                 14.782    -0.136               0.041  linear_regression

Based on the performance of the three best models on the validation set, we'd choose XGBoost as our final model. Of the three models, XGBoost has the lowest MAE. <br>
Additionally, it also has the lowest MSE, MSLE and Median Absolute Error.

# Final Best Model
Now, using the best model of the three models we compared, let's get its performance on the test set.

In [ ]:
%%bigquery xgb_model_test --project $project_name
SELECT
    *
FROM
    ML.PREDICT(MODEL `adl-analytics.return_prediction_ga4.xgb_model`,
        (
            SELECT
                * EXCEPT(user_pseudo_id, ecommerce_transaction_id, split, ecommerce_refund_value_in_usd),
                ecommerce_refund_value_in_usd AS label,
            FROM
                `return_prediction_ga4.step_6_merged_scaled`
            WHERE
                split = 'TEST'
        ),
            STRUCT(55 AS trial_id));

Query is running:   0%|          |

Downloading:   0%|          |

In [ ]:
# model predictions
xgb_model_test.head()

trial_id  predicted_label  transaction_ga_session_number  user_ltv_revenue  \
0        55            0.974                          0.328            -0.171   
1        55            0.782                          0.328            -0.572   
2        55            1.078                          0.328            -0.133   
3        55            3.125                          1.582             1.404   
4        55            1.390                          0.829             0.086   

   ecommerce_tax_value_in_usd  sum_item_price_in_usd  sum_item_promotions  \
0                      -0.576                 -0.703                0.091   
1                      -0.724                 -0.836               -0.516   
2                      -0.132                 -0.323               -0.516   
3                      -0.428                 -0.323                0.091   
4                       0.461                  0.470               -0.516   

   sum_item_quantity  sum_event_name_begin_checkout  \
0             -0.350                         -0.508   
1             -0.350                         -0.508   
2             -0.350                         -0.508   
3             -0.350                         -0.222   
4             -0.259                         -0.365   

   sum_event_name_select_item  sum_device_category_mobile  \
0                      -0.470                       1.209   
1                      -0.470                       1.209   
2                      -0.470                       1.209   
3                      -0.470                       1.209   
4                      -0.470                       1.209   

   sum_device_mobile_brand_name_Google  sum_device_mobile_brand_name_Huawei  \
0                               -0.740                               -0.138   
1                               -0.740                               -0.138   
2                               -0.740                               -0.138   
3                               -0.740                               -0.138   
4                               -0.740                               -0.138   

   sum_device_mobile_brand_name_Microsoft  \
0                                  -0.124   
1                                  -0.124   
2                                  -0.124   
3                                  -0.124   
4                                  -0.124   

   sum_device_mobile_brand_name_Samsung  sum_device_mobile_brand_name_Xiaomi  \
0                                -0.286                               -0.131   
1                                -0.286                               -0.131   
2                                -0.286                               -0.131   
3                                -0.286                               -0.131   
4                                -0.286                               -0.131   

   sum_device_mobile_model_name_ChromeBook  \
0                                   -0.240   
1                                   -0.240   
2                                   -0.240   
3                                   -0.240   
4                                   -0.240   

   sum_device_mobile_model_name_Safari  sum_device_mobile_model_name_iPad  \
0                               -0.510                             -0.102   
1                               -0.510                             -0.102   
2                               -0.510                             -0.102   
3                               -0.510                             -0.102   
4                               -0.510                             -0.102   

   sum_device_mobile_model_name_iPhone  \
0                                1.907   
1                                1.907   
2                                1.907   
3                                1.907   
4                                1.907   

   sum_device_web_info_browser_AndroidWebview  \
0                                      -0.125   
1                                    

### Evaluate model predictions

In [ ]:
# MAE
mae = mean_absolute_error(xgb_model_test['label'], xgb_model_test['predicted_label'])
mae

10.086418785968748

In [ ]:
# MSE
mse = mean_squared_error(xgb_model_test['label'], xgb_model_test['predicted_label'])
mse

896.4449986712499

In [ ]:
# MSLE
msle = mean_squared_log_error(xgb_model_test['label'], xgb_model_test['predicted_label'])
msle

2.103934618922581

In [ ]:
# Median Absolute Error
mmae = median_absolute_error(xgb_model_test['label'], xgb_model_test['predicted_label'])
mmae

1.3851211667060852

In [ ]:
# R2
r2 = r2_score(xgb_model_test['label'], xgb_model_test['predicted_label'])
r2

-0.055215712263408934

In [ ]:
# Explained Variance
ev = explained_variance_score(xgb_model_test['label'], xgb_model_test['predicted_label'])
ev

0.012488355839422538

In [ ]:
xgb_model_test_perf = pd.DataFrame([(55, mae, mse, msle, mmae, r2, ev)], columns=xgb_model_eval.columns)

In [ ]:
xgb_model_perf = pd.concat([xgb_model_eval, xgb_model_test_perf])
xgb_model_perf['dataset'] = ['VALIDATION', 'TEST']
xgb_model_perf

trial_id  mean_absolute_error  mean_squared_error  mean_squared_log_error  \
0        55               11.984             859.500                   2.277   
0        55               10.086             896.445                   2.104   

   median_absolute_error  r2_score  explained_variance     dataset  
0                  1.217    -0.099               0.025  VALIDATION  
0                  1.385    -0.055               0.012        TEST

Comparing the performance betwen validation and test, the model seems to have good generalizability since its performance on validation and test is similar.

# Feature Importance
If you are interested in knowing which features your model relied most heavily on to make predictions, you can use the option `ENABLE_GLOBAL_EXPLAIN = TRUE` when training your model with BQML.

# Conclusion